bentoml/_internal/utils/analytics/usage_stats.py

```diff
@@ -0,0 +1,170 @@
+import os
+import typing as t
+import logging
+import secrets
+import platform
+import threading
+from typing import TYPE_CHECKING
+from datetime import datetime
+from datetime import timezone
+from functools import wraps
+from functools import lru_cache
+
+import yaml
+import psutil
+import requests
+from simple_di import inject
+from simple_di import Provide
+
+from bentoml import __version__ as BENTOML_VERSION
+
+from ...configuration.containers import CLIENT_ID_PATH
+from ...configuration.containers import BentoMLContainer
+
+if TYPE_CHECKING:
+    P = t.ParamSpec("P")
+    T = t.TypeVar("T")
+    AsyncFunc = t.Callable[P, t.Coroutine[t.Any, t.Any, t.Any]]
+
+logger = logging.getLogger(__name__)
+
+BENTOML_DO_NOT_TRACK = "BENTOML_DO_NOT_TRACK"
+BENTOML_TRACKING_URL = "https://t.bentoml.com"
+BENTOML_USAGE_REPORT_INTERVAL_SECONDS = "__BENTOML_USAGE_REPORT_INTERVAL_SECONDS"
+BENTO_SERVE_SCHEDULED_TRACK_EVENT_TYPE = "bentoml_bento_serve_scheduled"
+
+
+def slient(func: "t.Callable[P, T]") -> "t.Callable[P, T]":  # pragma: no cover
+    # Slient errors when tracking
+    @wraps(func)
+    def wrapper(*args: "P.args", **kwargs: "P.kwargs") -> t.Any:
+        try:
+            return func(*args, **kwargs)
+        except Exception as err:  # pylint: disable=broad-except
+            logger.debug(f"{err}")
+
+    return wrapper
+
+
+@lru_cache(maxsize=1)
+def do_not_track() -> bool:
+    # Returns True if and only if the environment variable is defined and has value True.
+    # The function is cached for better performance.
+    return os.environ.get(BENTOML_DO_NOT_TRACK, str(False)).lower() == "true"
+
+
+@lru_cache(maxsize=1)
+def get_usage_stats_interval_seconds() -> int:
+    # The interval for getting usage stats for a serve is 12 hours.
+    # cached for better performance.
+    return int(os.environ.get(BENTOML_USAGE_REPORT_INTERVAL_SECONDS, 12 * 60 * 60))
```

Comment:
why do we need this env var?

Reply:
imo we should allow users to configure a period that they want to send the usage tracking if they want to do so